modules/nf-core/variancepartition/dream/templates/dream.R

```diff
@@ -0,0 +1,190 @@
+    method             = "ls",                    # Method for lmFit
+    proportion         = 0.01,                    # Proportion for eBayes
+    stdev_coef_lim     = "0.1,4",                 # Standard deviation coefficient limits for eBayes
+    trend              = FALSE,                   # Whether to use trend in eBayes
+    robust             = FALSE,                   # Whether to use robust method in eBayes
+    winsor_tail_p      = "0.05,0.1",              # Winsor tail probabilities for eBayes
+    ddf                = "adaptive",              # 'Satterthwaite', 'Kenward-Roger', or 'adaptive'
+    reml               = FALSE
+)
+
+# Load external arguments to opt list
+args_opt <- parse_args("$task.ext.args")
+for (ao in names(args_opt)) {
+    if (!ao %in% names(opt)) {
+        stop(paste("Invalid option:", ao))
+    }
+    opt[[ao]] <- args_opt[[ao]]
+}
+
+# If there are no blocking factors, convert string "null" to NULL
+if (!is.null(opt\$blocking_factors) && tolower(opt\$blocking_factors) == "null") {
+    opt\$blocking_factors <- NULL
+}
+
+# Load metadata
+metadata <- read_delim_flexible(opt\$sample_file, header = TRUE, stringsAsFactors = TRUE)
+rownames(metadata) <- metadata[[opt\$sample_id_col]]
+
+# Ensure contrast variable is factor, then relevel
+if (!is.null(opt\$contrast_reference) && opt\$contrast_reference != "") {
+    metadata[[opt\$contrast_variable]] <- factor(metadata[[opt\$contrast_variable]])
+    metadata[[opt\$contrast_variable]] <- relevel(metadata[[opt\$contrast_variable]], ref = opt\$contrast_reference)
+}
+
+# Exclude samples in metadata if specified
+if (!is.null(opt\$exclude_samples_col) && !is.null(opt\$exclude_samples_values)) {
+  metadata <- metadata[!(metadata[[opt\$exclude_samples_col]] %in% opt\$exclude_samples_values), ]
+}
+
+# Load count data
+countMatrix <- read_delim_flexible(opt\$count_file, header = TRUE, stringsAsFactors = FALSE)
+rownames(countMatrix) <- countMatrix\$gene_id # count_file/matrix must have a gene_id column.
+countMatrix <- countMatrix[, rownames(metadata), drop = FALSE]
+
+# Standard usage of limma/voom
+dge <- DGEList(countMatrix)
+dge <- calcNormFactors(dge)
+
+# Print sample names
+cat("Response sample names (dge counts):\n")
+print(colnames(dge\$counts))
+cat("Metadata sample names (rownames(metadata)):\n")
+print(rownames(metadata))
+
+# Construct model formula using contrast and, if available, blocking variables
+if (!is.null(opt\$blocking_factors) && opt\$blocking_factors != "") {
+  form <- as.formula(paste0("~ ", opt\$contrast_variable, " + (1 | ", opt\$blocking_factors, ")"))
+} else {
+  form <- as.formula(paste0("~ ", opt\$contrast_variable))
+}
+
+# Parallel processing setup
+threads <- as.numeric(opt\$threads)
+param <- SnowParam(threads, "SOCK", progressbar = TRUE)
+
+# Estimate weights using the linear mixed model of DREAM
+vobjDream <- voomWithDreamWeights(dge, form, metadata, BPPARAM = param)
```

Comment:
Can we make voom optional? DREAM can reasonably be applied to non-count data if it roughly follows a normal distribution.

Reply:
Sure. This was also part of the original `dev_tmp` implementation:
https://github.com/nf-core/differentialabundance/blob/f2c9346d1b6e773e8164f67856e32834f9dfd9e6/modules/nf-core/dream/differential/resources/usr/bin/dream_de.R#L391-L393

Reply:
Sure. This was also part of the original dev_tmp implementation:
https://github.com/nf-core/differentialabundance/blob/f2c9346d1b6e773e8164f67856e32834f9dfd9e6/modules/nf-core/dream/differential/resources/usr/bin/dream_de.R#L391-L393